hadoop-hdds/rocks-native/src/main/patches/rocks-native.patch

```diff
@@ -16,592 +16,557 @@
++      rep_->read_options_, rep_->moptions_.prefix_extractor.get(),
++      /*arena=*/nullptr, /*skip_filters=*/false,
++      TableReaderCaller::kSSTDumpTool);
++  return new RawSstFileReaderIterator(iter, has_from, from, has_to, to);
++
 +}
++}  // namespace ROCKSDB_NAMESPACE
++
++#endif  // ROCKSDB_LITE
+diff --git a/tools/raw_sst_file_reader_iterator.cc b/tools/raw_sst_file_reader_iterator.cc
+new file mode 100644
+index 000000000..39a2b84be
+--- /dev/null
++++ b/tools/raw_sst_file_reader_iterator.cc
+@@ -0,0 +1,94 @@
++//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
++//  This source code is licensed under both the GPLv2 (found in the
++//  COPYING file in the root directory) and Apache 2.0 License
++//  (found in the LICENSE.Apache file in the root directory).
++//
++#ifndef ROCKSDB_LITE
++
++#include "tools/raw_sst_file_reader_iterator.h"
++
++#include <chrono>
++#include <cinttypes>
++#include <iostream>
++#include <map>
++#include <memory>
++#include <sstream>
++#include <vector>
++
++#include "db/blob/blob_index.h"
++#include "db/memtable.h"
++#include "db/write_batch_internal.h"
++#include "options/cf_options.h"
++#include "port/port.h"
++#include "rocksdb/db.h"
++#include "rocksdb/env.h"
++#include "rocksdb/iterator.h"
++#include "rocksdb/slice_transform.h"
++#include "rocksdb/status.h"
++#include "rocksdb/table_properties.h"
++#include "rocksdb/utilities/ldb_cmd.h"
++#include "table/block_based/block.h"
++#include "table/block_based/block_based_table_builder.h"
++#include "table/block_based/block_based_table_factory.h"
++#include "table/block_based/block_builder.h"
++#include "table/format.h"
++#include "table/meta_blocks.h"
++#include "table/plain/plain_table_factory.h"
++#include "table/table_reader.h"
++#include "util/compression.h"
++#include "util/random.h"
++
++namespace ROCKSDB_NAMESPACE {
 +
-+void write(Slice &slice, FILE* file) {
-+  int size = (int)slice.size();
-+  write(slice.data(), size, file);
++RawSstFileReaderIterator::RawSstFileReaderIterator(InternalIterator* iterator,
++                             bool has_from, Slice* from_key,
++                             bool has_to, Slice* to_key)
++    : iter_(iterator),
++      ikey(new ParsedInternalKey()),
++      has_to_(has_to),
++      to_key_(to_key) {
++  if (has_from) {
++    InternalKey k;
++    k.SetMinPossibleForUserKey(*from_key);
++    iter_->Seek(k.Encode());
++  } else {
++    iter_->SeekToFirst();
++  }
++  initKey();
 +}
 +
-+void write(SequenceNumber sequenceNumber, FILE* file) {
++bool RawSstFileReaderIterator::has_next() const {
++  return iter_->Valid() && (!has_to_ ||
++                            BytewiseComparator()->Compare(
++                                getKey(), *to_key_) < 0);
++}
 +
-+  char b[8];
-+  int idx = 7;
-+  while (idx >= 0) {
-+    b[idx] = sequenceNumber % 256;
-+    sequenceNumber /= 256;
-+    idx -= 1;
++void RawSstFileReaderIterator::initKey() {
++  if (iter_->Valid()) {
```

Comment:
Does `iter_->Next();` throw an error if there is no more element? If not, we should throw an error if iterator is not valid.

Reply:
we are handling that on the java side. Let's keep it simple on the c++ side.

Reply:
I believe `iter_->Next();` would throw an exception.